lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -0,0 +1,580 @@
+          for (const auto &i3 : i2.second)
+            llvm::dbgs() << "\n node :" << i1.first << "," << i2.first
+                         << " is connected to ," << i3.val << "," << i3.fieldId;
+    });
+    return success();
+  }
+
+  void printError(Value node, ValueOpPair &childNode) {
+    if (!printCycle || currentPath.back() == node) {
+      printCycle = false;
+      return;
+    }
+    if (childNode.second)
+      TypeSwitch<Operation *>(childNode.second)
+          .Case<InstanceOp>([&](InstanceOp ins) {
+            size_t portNum = 0, childPortNum = 0;
+            // Get the port index for `node.val`.
+            for (const auto &p : llvm::enumerate(ins->getOpResults())) {
+              if (p.value() == childNode.first)
+                childPortNum = p.index();
+              if (p.value() == node)
+                portNum = p.index();
+            }
+            auto remark = ins.emitRemark("instance is part of a combinational "
+                                         "cycle, instance port number '")
+                          << childPortNum << "' has a path from port number '"
+                          << portNum << "'";
+            FieldRef fOut = getFieldRef(node);
+            FieldRef fIn = getFieldRef(childNode.first);
+            auto inName = getFieldName(fIn).first;
+            auto outName = getFieldName(fOut).first;
+            remark << ", " << inName << " <- " << outName;
+          })
+          .Case<MemOp>([&](MemOp mem) {
+            mem.emitRemark("memory is part of a combinational cycle ");
+          })
+          .Case<SubfieldOp, SubindexOp, SubaccessOp>([&](auto op) {})
+          .Default([&](auto owner) {
+            owner->emitRemark(
+                "this operation is part of the combinational cycle");
+          });
+  }
+
+  FModuleOp module;
+  InstanceGraph &instanceGraph;
+  // Set of visiting and visited nodes.
+  llvm::SmallDenseSet<Value> visiting, visited;
+  // Combinational paths between the ports of a module.
+  llvm::SmallDenseMap<Value, llvm::SmallDenseMap<size_t, SmallVector<Node>>>
+      &portPaths;
+  // Map of each unique storage location id and the list of SSA values that
+  // refer to it.
+  llvm::DenseMap<size_t, SmallVector<Value>> storageLocIdToValuesMap;
+  // Map of an SSA value to the list of storage location ids, that it can refer
+  // to.
+  llvm::DenseMap<Value, SmallVector<size_t>> valToStorageLocIdMap;
+  // For each leaf value, record the aggregate base value that it corresponds
+  // to.
+  llvm::DenseMap<Value, Value> leafToBaseValMap;
+  size_t storageLocId = 0;
+  SmallVector<Value> currentPath;
+  bool printCycle = true;
+  bool detectedCycle = false;
+};
+
+/// This pass constructs a local graph for each module to detect combinational
+/// cycles. To capture the cross-module combinational cycles, this pass inlines
+/// the combinational paths between IOs of its subinstances into a subgraph and
+/// encodes them in a `combPathsMap`.
```

Comment:
Update this comment, there's no `combPathsMap` in this version.

Reply:
Thanks, updated the map name.